engine/src/main/java/io/zeebe/engine/nwe/container/MultiInstanceBodyProcessor.java

```diff
@@ -0,0 +1,337 @@
+
+    if (updatedSuccessfully.isPresent() && !updatedSuccessfully.get()) {
+      // An incident was raised while updating the output collection, stop handling activity
+      return;
+    }
+
+    if (stateBehavior.isLastActiveExecutionPathInScope(childContext)) {
+      stateTransitionBehavior.transitionToCompleting(flowScopeContext);
+    }
+  }
+
+  @Override
+  public void onChildTerminated(
+      final ExecutableMultiInstanceBody element,
+      final BpmnElementContext flowScopeContext,
+      final BpmnElementContext childContext) {
+
+    if (flowScopeContext.getIntent() == WorkflowInstanceIntent.ELEMENT_TERMINATING
+        && stateBehavior.isLastActiveExecutionPathInScope(childContext)) {
+      stateTransitionBehavior.transitionToTerminated(flowScopeContext);
+
+    } else {
+      eventSubscriptionBehavior.publishTriggeredEventSubProcess(flowScopeContext);
+    }
+  }
+
+  private boolean updateOutputCollection(
+      final ExecutableMultiInstanceBody element,
+      final BpmnElementContext childContext,
+      final DirectBuffer variableName) {
+
+    final var bodyInstanceKey = childContext.getFlowScopeKey();
+    final var loopCounter =
+        stateBehavior.getElementInstance(childContext).getMultiInstanceLoopCounter();
+
+    final Optional<DirectBuffer> elementVariable = readOutputElementVariable(element, childContext);
+    if (elementVariable.isEmpty()) {
+      return false;
+    }
+
+    // we need to read the output element variable before the current collection is read,
+    // because readOutputElementVariable(Context) uses the same buffer as getVariableLocal
+    // this could also be avoided by cloning the current collection, but that is slower.
+    final var currentCollection = variablesState.getVariableLocal(bodyInstanceKey, variableName);
+    final var updatedCollection = insertAt(currentCollection, loopCounter, elementVariable.get());
+    variablesState.setVariableLocal(
+        bodyInstanceKey, childContext.getWorkflowKey(), variableName, updatedCollection);
+
+    return true;
+  }
+
+  private Optional<DirectBuffer> readOutputElementVariable(
+      final ExecutableMultiInstanceBody element, final BpmnElementContext context) {
+    final var expression = element.getLoopCharacteristics().getOutputElement().orElseThrow();
+    return expressionBehavior.evaluateAnyExpression(expression, context.toStepContext());
+  }
+
+  private DirectBuffer insertAt(
+      final DirectBuffer array, final int index, final DirectBuffer element) {
+
+    variableReader.wrap(array, 0, array.capacity());
+    variableReader.readArrayHeader();
+    variableReader.skipValues((long) index - 1L);
+
+    final var offsetBefore = variableReader.getOffset();
+    variableReader.skipValue();
+    final var offsetAfter = variableReader.getOffset();
+
+    variableWriter.wrap(variableBuffer, 0);
+    variableWriter.writeRaw(array, 0, offsetBefore);
```

Comment:
Why we need to rewrite the collection?

Reply:
It is the same logic as the current implementation. We need to insert an element into the collection. Feel free to suggest another approach (in a separate issue maybe).